File: 2-medium/00797-all-paths-from-source-to-target/solution.cpp
class Solution {
    void backtrack(vector<vector<int>>& graph, int curr_index, vector<int> curr_vec, vector<vector<int>> &res) {

        curr_vec.push_back(curr_index);

        if (curr_index == graph.size() - 1 ){
            res.push_back(curr_vec);
        } else {
            for (auto g: graph[curr_index]) {
                backtrack(graph, g, curr_vec, res);
            }
        }

        curr_vec.pop_back();
    }
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        vector<vector<int>> res{};

        backtrack(graph, 0, vector<int>(), res);

        return res;
    }
};

File: 2-medium/00797-all-paths-from-source-to-target/solution.rs
impl Solution {
    fn backtrack(graph: &Vec<Vec<i32>>, curr_index: usize, curr_vec: &mut Vec<i32>, res: &mut Vec<Vec<i32>>) {        

        curr_vec.push(curr_index as i32);

        if curr_index == graph.len() - 1 {
            res.push(curr_vec.clone());
        } else {
            for &g in graph[curr_index].iter() {
                Self::backtrack(graph, g as usize, curr_vec, res);
            }
        }

        curr_vec.pop();
    }
    pub fn all_paths_source_target(graph: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let mut res: Vec<Vec<i32>> = Vec::new();
        
        Self::backtrack(&graph, 0, &mut Vec::<i32>::new(), &mut res);

        res
    }
}
